### PR TITLE
[youtube] Add :ytnotifications extractor

### PR DESCRIPTION
**Pull request by krichbanana**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

When you enable notifications on a channel, videos will show in the notification menu (the "bell") when videos are uploaded or go live. This extractor adds an option to get these video URLs with just a simple API call. Community posts and video comment replies may show up as notifications, but they should be ignored. It may be possible to extract from those posts/replies, but that was beyond my original goal for this new feature.

Continuations are supported. Since notifications can be practically endless, the extractor keyword has a value argument that specifies the maximum number of pages to fetch (each page has 20 real entries), with a sane default of 1 (i.e. no continuation). I haven't tested to see how far back it goes for fear of throttling, but I can fetch 20 pages (~400 entries) in 4 seconds.

Since this is an unusual part of the API being used, some commonly used methods were ignored; I tried to reuse as much as I could.

The extractor creates a playlist of URLs with just a bit of info added (provided thumbnail, original title, uploader, and video id). The benefit of this extractor is that private videos still show up, and they include the uploader and video title.

Some explanation for excluded fields:
Dates are relative and imprecise (e.g. "1 day ago") and may also be plainly inaccurate. The action in the message text (e.g. "uploaded", "is live") is not accurate (a livestream vod that was unprivated long after the stream ended would show as "uploaded", for example).

The included thumbnail is weird: for livestreams only, the height is set to 0. There is also the issue that without flat-playlist, this thumbnail is removed from the infodict's list of thumbnails entirely.

A few of the methods could potentially be inlined.

TODO: document the feature.

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        # notificationRenderer
+        video_id = traverse_obj(
+            notification, ('navigationEndpoint', 'watchEndpoint', 'videoId'), expected_type=str)
+        if not video_id:
+            return None
+        thumbnails = traverse_obj(notification, ('videoThumbnail', 'thumbnails'), expected_type=list)
+        status_text = traverse_obj(notification, ('shortMessage', 'simpleText'), expected_type=str)
+        uploader = traverse_obj(
+            notification, ('contextualMenu', 'menuRenderer', 'items', 1, 'menuServiceItemRenderer', 'text', 'runs', 1, 'text'), expected_type=str)
+        status_text = status_text.removeprefix(uploader + ' ')
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5550]: removeprefix is 3.9+. Use `remove_start`

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        split_text = status_text.split(': ', 2)
+        title = None
+        if len(split_text) == 2:
+            title = split_text[1]
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5554]: can be simplified with a single regex

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        extra_info = {
+            'thumbnails': thumbnails,
+            'uploader': uploader,
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5558]: why even construct a dict?

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+                    if key == 'notificationRenderer':
+                        yield value
+                        break
+                    elif key == 'continuationItemRenderer':
+                        continuation_list[0] = value
+                        return
+        continuation_list[0] = None
+
+    def _extract_notification(self, notification):
+        # notificationRenderer
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5541]:
```suggestion
    def _extract_notification_renderer(self, notification):
```

File: yt_dlp/extractor/youtube.py [hunk 5]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        video_id = traverse_obj(
+            notification, ('navigationEndpoint', 'watchEndpoint', 'videoId'), expected_type=str)
+        if not video_id:
+            return None
+        thumbnails = traverse_obj(notification, ('videoThumbnail', 'thumbnails'), expected_type=list)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5546]: _extract_thumbnails

File: yt_dlp/extractor/youtube.py [hunk 6]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+            'thumbnails': thumbnails,
+            'uploader': uploader,
+        }
+        return self.url_result(
+            f'https://www.youtube.com/watch?v={video_id}', ie=YoutubeIE.ie_key(),
+            video_id=video_id, video_title=title, **extra_info)
+
+    @classmethod
+    def _extract_next_notification_continuation_data(cls, renderer):
+        # continuationItemRenderer
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5565]:
```suggestion
    @classmethod
    def _extract_next_notification_continuation_data(cls, renderer):
```

File: yt_dlp/extractor/youtube.py [hunk 7]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        if not ctoken:
+            return
+        return cls._build_notification_continuation_query(ctoken)
+
+    @staticmethod
+    def _build_notification_continuation_query(ctoken):
+        query = {
+            'ctoken': ctoken
+        }
+        return query
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5577]:
```suggestion
        return {
            'ctoken': ctoken
        } if ctoken else None
```

File: yt_dlp/extractor/youtube.py [hunk 8]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+
+    def _notification_menu_entries(self, headers, max_requests):
+        continuation_list = [None]
+        request_count = 0
+        entries = []
+        while request_count < max_requests:
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5583]:
```suggestion
        for page in itertools.count():
```

File: yt_dlp/extractor/youtube.py [hunk 9]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+            request_count += 1
+            response = self._extract_response(
+                note=f'Downloading API JSON (page {request_count})',
+                ep='notification/get_notification_menu', headers=headers,
+                check_get_keys=['responseContext', 'actions', 'trackingParams'],
+                item_id='notification_menu', query=query)
+            for notification in self._extract_notification_menu(response, continuation_list) or []:
+                entry = self._extract_notification(notification)
+                if entry:
+                    entries.append(entry)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5596]:
```suggestion
                    yield entry
```

File: yt_dlp/extractor/youtube.py [hunk 10]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        title = None
+        if len(split_text) == 2:
+            title = split_text[1]
+        extra_info = {
+            'thumbnails': thumbnails,
+            'uploader': uploader,
+        }
+        return self.url_result(
+            f'https://www.youtube.com/watch?v={video_id}', ie=YoutubeIE.ie_key(),
+            video_id=video_id, video_title=title, **extra_info)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5561]:
```suggestion
        return self.url_result(
            f'https://www.youtube.com/watch?v={video_id}', YoutubeIE,
            video_id, title, thumbnails=thumbnails, uploader=uploader)
```
or, if we end up extracting more fields, it's better to do:
```py
        return {
            '_type': 'url',
            'url': f'https://www.youtube.com/watch?v={video_id}',
            'ie_key': YoutubeIE.ie_key(),
            'video_id': video_id,
            'title': title,
            ...
        }
```

File: yt_dlp/extractor/youtube.py [hunk 11]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+
+    @classmethod
+    def _extract_next_notification_continuation_data(cls, renderer):
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5564]: I personally don't like creating class/staticmethods without any reason. But 🤷

File: yt_dlp/extractor/youtube.py [hunk 12]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+            query = {}
+            if continuation_list[0]:
+                query = self._extract_next_notification_continuation_data(continuation_list[0])
+            request_count += 1
+            response = self._extract_response(
+                note=f'Downloading API JSON (page {request_count})',
+                ep='notification/get_notification_menu', headers=headers,
+                check_get_keys=['responseContext', 'actions', 'trackingParams'],
+                item_id='notification_menu', query=query)
+            for notification in self._extract_notification_menu(response, continuation_list) or []:
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5593]: generators are always iterable
```suggestion
            for notification in self._extract_notification_menu(response, continuation_list):
```

File: yt_dlp/extractor/youtube.py [hunk 13]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+                entry = self._extract_notification(notification)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5594]:
```suggestion
            for entry in map(self._extract_notification, self._extract_notification_menu(response, continuation_list)):
```
or even better, in `_extract_notification_menu`, directly `yield self._extract_notification(renderer)`

File: yt_dlp/extractor/youtube.py [hunk 14]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+                if entry:
+                    entries.append(entry)
+            if not continuation_list[0]:
+                break
+        return entries
+
+    def get_notifications(self, max_requests=1):
+        return self._notification_menu_entries(self.generate_api_headers(), max_requests)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5602]: Why not just generate the headers inside `_notification_menu_entries`? Then this function isn't even needed

File: yt_dlp/extractor/youtube.py [hunk 15]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+
+    def _real_extract(self, url):
+        max_pages = int(self._match_valid_url(url).group('max_pages') or 1)
+        self.write_debug(f'max_pages: {max_pages}')
+        return self.playlist_result(
+            entries=self.get_notifications(max_requests=max_pages), playlist_title='Notifications', ie=YoutubeTabIE.ie_key())
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5608]:
```suggestion
            entries=self.get_notifications(), playlist_title='Youtube Notifications')
```
You can't set `ie` for playlist!

**Review by pukkandan — CHANGES_REQUESTED** [review]
> Community posts and video comment replies may show up as notifications, but they should be ignored. It may be possible to extract from those posts/replies, but that was beyond my original goal for this new feature.

Nothing to be done about comments, but we could extract posts since posts can contain embedded videos

> Since notifications can be practically endless, the extractor keyword has a value argument that specifies the maximum number of pages to fetch (each page has 20 real entries), with a sane default of 1 (i.e. no continuation). I haven't tested to see how far back it goes for fear of throttling, but I can fetch 20 pages (~400 entries) in 4 seconds.

No need for any artificial limitations/extra parameters. As long as the extractor yields the results lazily, `--playlist-end` will handle the rest

> Dates are relative and imprecise (e.g. "1 day ago") and may also be plainly inaccurate. The action in the message text (e.g. "uploaded", "is live") is not accurate (a livestream vod that was unprivated long after the stream ended would show as "uploaded", for example).

We generally do not want to return inaccurate fields. But For `upload_date`, obey the `approximate_date` extractor-arg

> The included thumbnail is weird: for livestreams only, the height is set to 0. 

We should remove the empty thumbnail imo. But returning it is technically fine

> There is also the issue that without flat-playlist, this thumbnail is removed from the infodict's list of thumbnails entirely.

That is expected behavior. The thumbnails we get from video extraction is more accurate information


PS: Some of my review comments are inter-related. Pls read all of them before starting to make changes

File: yt_dlp/extractor/youtube.py [hunk 16]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+
+    def _extract_notification(self, notification):
+        # notificationRenderer
+        video_id = traverse_obj(
+            notification, ('navigationEndpoint', 'watchEndpoint', 'videoId'), expected_type=str)
+        if not video_id:
+            return None
+        thumbnails = traverse_obj(notification, ('videoThumbnail', 'thumbnails'), expected_type=list)
+        status_text = traverse_obj(notification, ('shortMessage', 'simpleText'), expected_type=str)
+        uploader = traverse_obj(
```
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:5548]: there's also `_get_text` that's similar to traverse_obj but handles youtube's simpletext/runs

File: yt_dlp/extractor/youtube.py [hunk 17]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        entries = []
+        while request_count < max_requests:
+            query = {}
+            if continuation_list[0]:
+                query = self._extract_next_notification_continuation_data(continuation_list[0])
+            request_count += 1
+            response = self._extract_response(
+                note=f'Downloading API JSON (page {request_count})',
+                ep='notification/get_notification_menu', headers=headers,
+                check_get_keys=['responseContext', 'actions', 'trackingParams'],
```
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:5591]: only need `actions` if that is all you are accessing

**Review by coletdjnz — COMMENTED** [review]
It might be possible we can merge some of this into YoutubeTabBaseIE

when I get some time I'll have a look to see what is feasible to do in this PR (i.e., doesn't require any large refactoring to existing code)

File: yt_dlp/extractor/youtube.py (hunk 11 above)
Comment (krichbanana) [reply yt_dlp/extractor/youtube.py:5563]: I was trying to match the style elsewhere, but fair enough.

File: yt_dlp/extractor/youtube.py [hunk 18]
```diff
@@ -5505,6 +5505,109 @@ def _real_extract(self, url):
...
+        'only_matching': True,
+    }]
+
+    @staticmethod
+    def _extract_notification_menu(response, continuation_list):
+        notification_list = traverse_obj(
+            response, ('actions', 0, 'openPopupAction', 'popup', 'multiPageMenuRenderer', 'sections', 0, 'multiPageMenuNotificationSectionRenderer', 'items'), expected_type=list)
+        if not notification_list:
+            notification_list = traverse_obj(
+                response, ('actions', 0, 'appendContinuationItemsAction', 'continuationItems'), expected_type=list)
```
[the remark this answers is not in the document]
Comment (krichbanana) [reply yt_dlp/extractor/youtube.py]: It's size 1, and the ... adds an annoying list around the result.

**Comment by krichbanana** [comment]
I ended up adding posts. Posts are processed by youtubetab as a playlist and get influenced by --playlist-(start/end/items), so if you're trying to not download the first video the post won't get downloaded at all. I'm not sure if there is a workaround for that or if it's a general issue.

I noticed that the titles extracted from the video notifications had soft hyphens in them; I tried to remove them to match what the player shows, but I don't know if there is a more convenient method to do so.

**Comment by pukkandan** [comment]
> Posts are processed by youtubetab as a playlist and get influenced by --playlist-(start/end/items), so if you're trying to not download the first video of the notifications the post video won't get downloaded at all. I'm not sure if there is a workaround for that or if it's a general issue.

This is a general issue. Support for nested playlists is quite limited in yt-dlp

> I noticed that the titles extracted from the video notifications had soft hyphens in them; I tried to remove them to match what the player shows, but I don't know if there is a more convenient method to do so.

None of my notifications have them. Can you give an example of what the `status_text` (with hyphen) looks like?

---

I have pushed some changes. I don't have any posts in my notifications. So pls double check them. Videos are working perfectly (and comments are skipped)

Can be merged once @coletdjnz gives okay

**Review by pukkandan — APPROVED** [review]

**Comment by krichbanana** [comment]
> > I noticed that the titles extracted from the video notifications had soft hyphens in them; I tried to remove them to match what the player shows, but I don't know if there is a more convenient method to do so.
> 
> None of my notifications have them. Can you give an example of what the `status_text` (with hyphen) looks like?

```json
"shortMessage": {
    "simpleText": "Ceres Fauna Ch. hololive-EN premiering now: 【UNARCHIVED KARAOKE】 Fauna's DefinitelyMaybeTotal­lyNot Prerecorded Karaoke"
 },
```
It's not visible normally but shows up on a terminal; it's right before the "lyNot". Unprintable characters like these tend to make string matching difficult.

-----
> I have pushed some changes. I don't have any posts in my notifications. So pls double check them. Videos are working perfectly (and comments are skipped)

Title for member posts was set before this change to a short text preview of the post (now it's null). It wasn't the actual title of the video, but it may be nice to have something for a title.

For example, it'll show up as
```json
"For members only from Ceres Fauna Ch. hololive-EN: \"Next members' stream is tomorrow! <redacted...> It's ...\""
```
so mostly useless.

**Comment by coletdjnz** [comment]
There is an issue with auth with secondary accounts & channels.
We need some sort of webpage download to extract ytcfg

**Comment by coletdjnz** [comment]
- Fixed the auth issue - I've made a general function to download a client ytcfg. Supports `youtubetab:skip=webpage,authcheck` extractor args.

- Made some changes that may be slightly out of scope of this pr (adding same ytcfg download to search IE) - may want to commit that separately. 

- I saw pukkandan added `responseContext` to check_get_keys - I don't think we needed that. Not that either way does anything, it should always be there.

@pukkandan if you can give this a quick check

In the future a lot of this can probably be merged into TabIE. Currently it is not in a state we can do so easily.

**Review by coletdjnz — APPROVED** [review]

**Pull request merged**